mlkem/verify.h

```diff
@@ -9,66 +9,227 @@
+// to int16_t -> uint16_t.
+#ifdef CBMC
+#pragma CPROVER check push
+#pragma CPROVER check disable "conversion"
+#endif
+
+/*************************************************
+ * Name:        ct_sel_int16
+ *
+ * Description: Functionally equivalent to cond ? a : b,
+ *              but implemented with guards against
+ *              compiler-introduced branches.
+ *
+ * Arguments:   int16_t a:       First alternative
+ *              int16_t b:       Second alternative
+ *              uint16_t cond:   Condition variable.
+ **************************************************/
+STATIC_INLINE_TESTABLE int16_t ct_sel_int16(int16_t a, int16_t b, uint16_t cond)
+__contract__(ensures(return_value == (cond ? a : b)))
+{
+  uint16_t au = a, bu = b;
+  uint16_t res = bu ^ (ct_cmask_nonzero_u16(cond) & (au ^ bu));
+  return (int16_t)res;
+}
+
+// Put unsigned-to-signed warnings in CBMC back into scope
+#ifdef CBMC
+#pragma CPROVER check pop
+#endif
+
 /*************************************************
- * Name:        verify
+ * Name:        ct_sel_uint8
+ *
+ * Description: Functionally equivalent to cond ? a : b,
+ *              but implemented with guards against
+ *              compiler-introduced branches.
+ *
+ * Arguments:   uint8_t a:       First alternative
+ *              uint8_t b:       Second alternative
+ *              uuint8_t cond:   Condition variable.
+ **************************************************/
+STATIC_INLINE_TESTABLE uint8_t ct_sel_uint8(uint8_t a, uint8_t b, uint8_t cond)
+__contract__(ensures(return_value == (cond ? a : b)))
+{
+  return b ^ (ct_cmask_nonzero_u8(cond) & (a ^ b));
+}
+
+#define ct_memcmp MLKEM_NAMESPACE(ct_memcmp)
+/*************************************************
+ * Name:        ct_memcmp
  *
  * Description: Compare two arrays for equality in constant time.
  *
  * Arguments:   const uint8_t *a: pointer to first byte array
  *              const uint8_t *b: pointer to second byte array
  *              size_t len:       length of the byte arrays
  *
- * Returns 0 if the byte arrays are equal, 1 otherwise
+ * Returns 0 if the byte arrays are equal, a non-zero value otherwise
  **************************************************/
-int verify(const uint8_t *a, const uint8_t *b, const size_t len)
+STATIC_INLINE_TESTABLE uint8_t ct_memcmp(const uint8_t *a, const uint8_t *b,
+                                         const size_t len)
 __contract__(
   requires(memory_no_alias(a, len))
   requires(memory_no_alias(b, len))
   requires(len <= INT_MAX)
-  ensures(return_value == (1 - forall(int, i, 0, ((int)len - 1), (a[i] == b[i]))))
-);
+  ensures((return_value == 0) == forall(int, i, 0, ((int)len - 1), (a[i] == b[i]))))
+{
+  uint8_t r = 0, s = 0;
+
+  // Switch to a _signed_ ilen value, so that our loop counter
+  // can also be signed, and thus (i - 1) in the loop invariant
+  // can yield -1 as required.
+  const int ilen = (int)len;
+
+  for (int i = 0; i < ilen; i++)
```

Comment:
In the following function you just used size_t for the loop counter as well. 
Did you try that here, too?

Reply:
This is because we have a more complicated loop invariant which is easier to state with `int` than `size_t`.

Reply:
okay.